src/stream/src/common/log_store/kv_log_store/buffer.rs

```diff
@@ -321,20 +311,67 @@ impl LogStoreBufferReceiver {
         }
     }
 
-    pub(crate) fn truncate(&mut self) {
+    pub(crate) fn truncate(&mut self, offset: TruncateOffset) {
         let mut inner = self.buffer.inner();
-        if let Some((epoch, item)) = inner.consumed_queue.front() {
+        let mut latest_offset: Option<ReaderTruncationOffsetType> = None;
+        while let Some((epoch, item)) = inner.consumed_queue.back() {
+            let epoch = *epoch;
             match item {
+                LogStoreBufferItem::StreamChunk {
+                    chunk_id,
+                    flushed,
+                    end_seq_id,
+                    ..
+                } => {
+                    let chunk_offset = TruncateOffset::Chunk {
+                        epoch,
+                        chunk_id: *chunk_id,
+                    };
+                    let flushed = *flushed;
+                    let end_seq_id = *end_seq_id;
+                    if chunk_offset <= offset {
+                        inner.stream_chunk_count -= 1;
+                        inner.consumed_queue.pop_back();
+                        if flushed {
+                            latest_offset = Some((epoch, Some(end_seq_id)));
+                        }
+                    } else {
+                        break;
+                    }
+                }
+                LogStoreBufferItem::Flushed {
+                    chunk_id,
+                    end_seq_id,
+                    ..
+                } => {
+                    let end_seq_id = *end_seq_id;
+                    let chunk_offset = TruncateOffset::Chunk {
+                        epoch,
+                        chunk_id: *chunk_id,
+                    };
+                    if chunk_offset <= offset {
+                        inner.consumed_queue.pop_back();
+                        latest_offset = Some((epoch, Some(end_seq_id)));
+                    } else {
+                        break;
+                    }
+                }
                 LogStoreBufferItem::Barrier { .. } => {
-                    inner.updated_truncation = Some(*epoch);
+                    let chunk_offset = TruncateOffset::Barrier { epoch };
+                    if chunk_offset <= offset {
+                        inner.consumed_queue.pop_back();
+                        latest_offset = Some((epoch, None));
+                    } else {
+                        break;
+                    }
                 }
-                _ => {
-                    unreachable!("should only call truncate right after getting a barrier");
+                LogStoreBufferItem::UpdateVnodes(_) => {
+                    inner.consumed_queue.pop_back();
                 }
             }
-            inner.consumed_queue.clear();
-            inner.stream_chunk_count -= inner.consumed_stream_chunk_count;
-            inner.consumed_stream_chunk_count = 0;
+        }
+        if let Some(offset) = latest_offset {
+            inner.updated_truncation = Some(offset);
```

Comment:
We use the largest flushed offset <= the provided truncate offset for truncation. Will there be an issue if we blindly use the provided truncate offset for truncation?

Reply:
The provided truncate offset is truncate at a provided chunk id, while in the state store we will truncate at seq id. The code above is to turn the chunk id offset into the seq id offset.